metagraph/src/cli/load/load_annotated_graph.cpp

```diff
@@ -41,39 +41,27 @@ std::unique_ptr<AnnotatedDBG> initialize_annotated_dbg(std::shared_ptr<DeBruijnG
         }
-        const Config::AnnotationType input_anno_type
-                = parse_annotation_type(config.infbase_annotators.at(0));
         // row_diff annotation is special, as it must know the graph structure
-        if (input_anno_type == Config::AnnotationType::RowDiff
-            || input_anno_type == Config::AnnotationType::RowDiffBRWT
-            || input_anno_type == Config::AnnotationType::RowDiffRowSparse) {
-            auto dbg_graph = dynamic_cast<const DBGSuccinct *>(graph.get());
-
-            if (!dbg_graph) {
-                if (auto *canonical = dynamic_cast<CanonicalDBG *>(graph.get()))
-                    dbg_graph = dynamic_cast<const DBGSuccinct *>(&canonical->get_graph());
+        using namespace annot::binmat;
+        BinaryMatrix &matrix = const_cast<BinaryMatrix &>(annotation_temp->get_matrix());
+        if (IRowDiff *row_diff = dynamic_cast<IRowDiff*>(&matrix)) {
+            const DBGSuccinct *dbg_graph;
+            if (auto *canonical = dynamic_cast<CanonicalDBG *>(graph.get())) {
+                dbg_graph = dynamic_cast<const DBGSuccinct *>(&canonical->get_graph());
+            } else {
+                dbg_graph = dynamic_cast<const DBGSuccinct *>(graph.get());
             }
-
             if (!dbg_graph) {
-                logger->error(
-                        "Only succinct de Bruijn graph representations are "
-                        "supported for row-diff annotations");
+                logger->error("Only succinct de Bruijn graph representations"
+                              " are supported for row-diff annotations");
                 std::exit(1);
             }
-            // this is really ugly, but the alternative is to add a set_graph method to
-            // all annotations
-            using namespace annot::binmat;
-            BinaryMatrix &matrix = const_cast<BinaryMatrix &>(annotation_temp->get_matrix());
-            if (input_anno_type == Config::AnnotationType::RowDiff) {
-                dynamic_cast<RowDiff<ColumnMajor> &>(matrix).set_graph(dbg_graph);
-                std::string anchor_fname = config.infbase + kRowDiffAnchorExt;
-                dynamic_cast<RowDiff<ColumnMajor> &>(matrix).load_anchor(anchor_fname);
-                std::string fork_succ_fname = config.infbase + kRowDiffForkSuccExt;
-                dynamic_cast<RowDiff<ColumnMajor> &>(matrix).load_fork_succ(fork_succ_fname);
-            } else if (input_anno_type == Config::AnnotationType::RowDiffRowSparse) {
-                dynamic_cast<RowDiff<RowSparse> &>(matrix).set_graph(dbg_graph);
-            } else {
-                dynamic_cast<RowDiff<BRWT> &>(matrix).set_graph(dbg_graph);
+
+            row_diff->set_graph(dbg_graph);
+
+            if (auto *row_diff_column = dynamic_cast<RowDiff<ColumnMajor> *>(&matrix)) {
+                row_diff_column->load_anchor(config.infbase + kRowDiffAnchorExt);
+                row_diff_column->load_fork_succ(config.infbase + kRowDiffForkSuccExt);
```

Comment:
as stated above, why not move this logic into `RowDiff::load`?

Reply:
This would require a different signature for `load`: `load(ann_fname, graph_fname)`. In contrast, all other annotations have `load(ann_fname)`. We can overload `load(ann_fname, graph_fname)` specially for the row-diff classes, but that will be kind of the same as this, with `load` + `load_anchor` + `load_fork_succ`